catkin_tools/jobs/utils.py

```diff
@@ -50,42 +50,67 @@ def get_env_loaders(package, context):
     return sources
 
 
+def merge_envs(job_env, overlay_envs):
+    '''
+    In the merged/linked case of single env, this function amounts to a straight
+    assignment, but a more complex merge is required with isolated result spaces,
+    since a package's build environment may require extending that of multiple
+    other result spaces.
+    '''
+    merge_path_values = {}
+
+    for overlay_env in overlay_envs:
+        for key, values_str in overlay_env.items():
+            if key.endswith('PATH'):
```

Comment:
In the implementation as currently proposed, the function _is_ also getting invoked in the non-isolated case. It's not really doing anything, but it's there ready to trip up a user in the scenario you describe— even without having a colon in the variable, you could get behaviour like:

1. Package installed in the underlay sets `FOO_PATH=bar`
2. Package in the current merged/linked resultspace overrides that to `FOO_PATH=baz`
3. Environment setup for next package build merges those into `FOO_PATH=baz:bar`, when the intended behaviour was a straight up override, not a prepend.

Hence the suggestion of either a whitelist of keys where prepending is _known_ to be correct behaviour, or else disabling this function altogether for the non-isolated case.